### PR TITLE
1022 wrong compiler version in ci for gcc

### LINKED ISSUE #1022: wrong compiler version in CI for gcc
### Bug description

gcc/gxx-11 gets installed, while only version 9 and 10 are used for building and testing

### Version

Github (CI, Actions, etc.)

### To reproduce

Bug found in .github/actions/linux-build/action.yml

### Relevant log output

_No response_

### Add any relevant information, e.g. used compiler, screenshots.

_No response_

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project

### PR DESCRIPTION
**Pull request by MaxBetzDLR**
# Changes and Information

Please **briefly list the changes** (main added features, changed items, or corrected bugs) made:

- Use gcc/gxx-11 in CI
- Update Readme

## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/SciCompMod/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/SciCompMod/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful (with and without OpenMP)
- [ ] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [ ] Proper attention to licenses, especially no new third-party software with conflicting license has been added
- [ ] (For ABM development) Checked [benchmark results](https://github.com/SciCompMod/memilio/wiki/Agent-Based-Model-Development) and ran and posted a local test above from before and after development to ensure performance is monitored.

### Checks by code reviewer(s)

- [x] Corresponding issue(s) is/are linked and addressed
- [x] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [x] Appropriate **unit tests** have been added, CI passes, code coverage and performance is acceptable (did not decrease)
- [x] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] On merge, add 2-5 lines with the changes (main added features, changed items, or corrected bugs) to the merge-commit-message. This can be taken from the **briefly-list-the-changes** above (best case) or the separate commit messages (worst case).

Closes #1022

### CODE REVIEW

File: cpp/README.md [hunk 1]
```diff
@@ -16,9 +16,9 @@ Directory structure:
 MEmilio C++ uses CMake as a build configuration system (https://cmake.org/)
 
 MEmilio C++ is regularly tested with the following compilers (list will be extended over time):
-- GCC, versions 7.3.0 - 10.2.0
-- Clang, version 9.0
-- MSVC, versions 19.16.27045.0 (Visual Studio 2017) - 19.29.30133.0 (Visual Studio 2019)
+- GCC, versions 9.4.0 - 11.4.0
+- Clang, version 6.0.1 - 14.0.0
+- MSVC, versions 19.29.30154.0 (Visual Studio 2019) - 19.38.33135.0 (Visual Studio 2022)
```
Comment (reneSchm) [inline cpp/README.md:21]:
```suggestion
- GCC, versions 9.4.0 and 11.4.0
- Clang, version 6.0.1 and 14.0.0
- MSVC, versions 19.29.30154.0 (Visual Studio 2019) and 19.38.33135.0 (Visual Studio 2022)
```
The dash looks like it indicates a range, I think we should change it (a comma would also be fine).

File: .github/actions/linux-build/action.yml [hunk 2]
```diff
@@ -73,8 +77,8 @@ runs:
             export CC=/usr/bin/gcc-9
             export CXX=/usr/bin/g++-9
           else        
-            export CC=/usr/bin/gcc-10
-            export CXX=/usr/bin/g++-10
+            export CC=/usr/bin/gcc-11
+            export CXX=/usr/bin/g++-11
```
Comment (reneSchm) [inline .github/actions/linux-build/action.yml:81]: Should we use variables for the version numbers now, or do we have a follow up issue for that?

**Review by reneSchm — CHANGES_REQUESTED** [review]
Just a minor change in the Readme, the rest looks good to me.

The change in abm_braunchschweig makes a lot of sense - I just removed the save_result_result in another branch as it was effectively unused, but stopping the entire ensemble because a singular output failed is likely not desirable. This here is better.

File: .github/actions/linux-build/action.yml (hunk 2 above)
Comment (mknaranja) [reply .github/actions/linux-build/action.yml:81]: I think @MaxBetzDLR already did that. Didn't you show me that?

File: cpp/README.md (hunk 1 above)
Comment (mknaranja) [reply cpp/README.md:21]: Is it ensured that the subversions do always stay the same? Maybe not? We only ask for gcc-11 for example. Maybe we should write:

`GCC, versions 9 and 11 (versions 9.4.0 and 11.4.0 as of May 07, 2024)`
?

File: cpp/simulations/abm_braunschweig.cpp [hunk 3]
```diff
@@ -1022,7 +1022,7 @@ mio::IOResult<void> run(const std::string& input_file, const fs::path& result_di
         // Option to save the current run result to file
         if (save_result_result && save_single_runs) {
             auto result_dir_run = result_dir / ("abm_result_run_" + std::to_string(run_idx) + ".h5");
-            BOOST_OUTCOME_TRY(save_result(ensemble_results.back(), loc_ids, 1, result_dir_run.string()));
+            save_result_result = save_result(ensemble_results.back(), loc_ids, 1, result_dir_run.string());
```
Comment (mknaranja) [inline cpp/simulations/abm_braunschweig.cpp:1025]: Why was this necessary? Looks like a nonintended (wrongly placed?) change here.

File: cpp/README.md (hunk 1 above)
Comment (reneSchm) [reply cpp/README.md:21]: Good point, I did not consider the subversions.

Though, I don't think they matter too much, it might be fine to just omit them, since they can get outdated quickly. There were two minor msvc updates I noticed recently, and I don't even have it installed.

File: .github/actions/linux-build/action.yml (hunk 2 above)
Comment (reneSchm) [reply .github/actions/linux-build/action.yml:81]: I see neither variables nor an issue related to this, what are you referring to?

File: cpp/simulations/abm_braunschweig.cpp (hunk 3 above)
Comment (reneSchm) [reply cpp/simulations/abm_braunschweig.cpp:1025]: The old version does produce a warning that save_result_result is uninitialized in the BOOST_OUTCOME_TRY a few lines below. Hence
> The change in abm_braunchschweig makes a lot of sense - I just removed the save_result_result in another branch as it was effectively unused, but stopping the entire ensemble because a singular output failed is likely not desirable. This here is better.

File: .github/actions/linux-build/action.yml (hunk 2 above)
Comment (mknaranja) [reply .github/actions/linux-build/action.yml:81]: Talk to him please.

File: cpp/README.md (hunk 1 above)
Comment (mknaranja) [reply cpp/README.md:21]: I think it makes sense to name at least the first number after the dot. There are sometimes significant changes from X.Y to X.Y+1

File: .github/actions/linux-build/action.yml (hunk 2 above)
Comment (MaxBetzDLR) [reply .github/actions/linux-build/action.yml:81]: Git actions do not support a good way for variables. The possible implementations do not seem sensible to me.

**Review by reneSchm — APPROVED** [review]

**Review by mknaranja — APPROVED** [review]
